src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -34470,6 +34470,269 @@ static void F(A<object?>[] c)
                 Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "a2.F").WithLocation(15, 13));
         }
 
+        [Fact]
+        public void ForEach_11()
+        {
+            var source =
+@"using System.Collections.Generic;
+class A
+{
+    public static implicit operator B?(A a) => null;
+}
+class B
+{
+}
+class C
+{
+    static void F(IEnumerable<A> e)
+    {
+        foreach (var x in e)
+            x.ToString();
+        foreach (B y in e)
+            y.ToString();
+        foreach (B? z in e)
+            z.ToString();
+    }
+}";
+            var comp = CreateCompilation(new[] { source, NonNullTypesTrue, NonNullTypesAttributesDefinition });
+            comp.VerifyDiagnostics(
+                // (15,18): warning CS8601: Possible null reference assignment.
```

Comment:
This error message is especially unhelpful in this context, as there's no assignment happening. Should we add a better message?

Reply:
Added comment.

---
In reply to: [214179736](https://github.com/dotnet/roslyn/pull/29587#discussion_r214179736) [](ancestors = 214179736)